content/post/2019-01-18-nerves-at-home-desk-controller.md

```diff
@@ -0,0 +1,245 @@
+most likely short for `Tx`/`Rx` and would be the communication with the controller,
+so I skipped those for now. Here's the final pinout I came up with:
+
+![desk_keypad_pinout](/images/2019-01-18/desk_keypad_pinout.png)
+
+Moving the desk is simply just connecting the yellow or green wire to GND. So that
+can be easily done by toggling a GPIO pin between _HIGH (1)_ and _LOW (0)_ using
+[Circuits.GPIO](https://github.com/elixir-circuits/circuits_gpio).
+
+```elixir
+# setting LOW will move the desk in whatever direction the pin is for
+# Be sure to initialize with HIGH to prevent moving on pin open.
+{:ok, up_pin} = Circuits.GPIO.open(23, :output, initial_value: 1)
+{:ok, down_pin} = Circuits.GPIO.open(24, :output, initial_value: 1)
+
+# set LOW to move up and back down
+Circuits.GPIO.write(up_pin, 0) # start moving
+:timer.sleep(500) 
+Circuits.GPIO.write(up_pin, 1) # stop moving
+
+Circuits.GPIO.write(down_pin, 0) # start moving
+:timer.sleep(500) 
+Circuits.GPIO.write(down_pin, 1) # stop moving
+```
+
+## Reading the messages
+
+Now comes the fun part and ultimate goal of this project: reading the messages 
+from the controller for the current height.
+
+Initially, I connected the `Tx`(white) and `Rx`(brown) wires to my computer via
+[USB to TTL serial cable](https://www.adafruit.com/product/954) I had around and
+used `screen` in my terminal to see what these messages would even look like.
+It was just jibberish and repeating characters, but at least there were messages!
+I tried lots of different baud rates thinking maybe thats was causing the garbled
+mess, but I'm not really proficient with `screen` and serial communication so
+I gave up there.
+
+Instead I decided to connect the wires to the corresponding UART pins on my rpi3
+and used yet another fancy library, [Circuits.UART](https://github.com/elixir-circuits/circuits_gpio).
+From there, I could open UART port and see if I get better messages there.
+
+```elixir
+{:ok, uart} = Circuits.UART.start_link
+# Open UART to send messages to current process
+:ok = Circuits.UART.open(uart, "ttyAMA0")
+
+# Move the desk a little bit and flush the messages
+flush
+```
+
+This exploded with messages! 
+```elixir
+{:circuits_uart, "ttyAMA0", <<255>>}
+{:circuits_uart, "ttyAMA0", <<1>>}
+{:circuits_uart, "ttyAMA0", <<1>>}
+{:circuits_uart, "ttyAMA0", <<0>>}
+{:circuits_uart, "ttyAMA0", <<253>>}
+{:circuits_uart, "ttyAMA0", <<1>>}
+{:circuits_uart, "ttyAMA0", <<1>>}
+{:circuits_uart, "ttyAMA0", <<0>>}
+{:circuits_uart, "ttyAMA0", <<254>>}
+{:circuits_uart, "ttyAMA0", <<1>>}
+{:circuits_uart, "ttyAMA0", <<5>>}
+{:circuits_uart, "ttyAMA0", <<1>>}
+{:circuits_uart, "ttyAMA0", <<0>>}
+...# repeated x100 
+```
+
+It was a bit overwelming and confusing. There were hundreds of messages for just
+simple movements and I was having trouble finding a pattern. I ended up pinging
+the always helpful folks in the [nerves slack channel](https://elixir-lang.slack.com/messages/C0AB4A879/)
```

Comment:
```suggestion
the always helpful folks in the [Nerves slack channel](https://elixir-lang.slack.com/messages/C0AB4A879/)
```

Reply:
Also, can you also add a link to the forum here?

Reply:
Done